### PR TITLE
feat(@angular/cli): improve unit test performance

### LINKED ISSUE #5423: Test development cycle is slow unless sourcemaps are turned off
The iterating unit test development is painfully slow (and appears to grow slower during a session) when running `ng test`. 

The console shows re-build 92% complete and then hangs for up to a minute (it seems anyway) before finally reporting test progress and refreshing the karma browser.

The test progress reporter ticks along at a slow pace. 

However, `ng test --sourcemap=false` (turning off sourcemaps) results in crisp iterations, with rebuild times in the sub-second range.

I'm describing my experience on a small app (the new Angular DocViewer) with fewer than 100 tests so far.

### Bug Report or Feature Request (mark with an `x`)
- [ ] bug report -> please search issues before submitting
- [x] feature request

### Versions.
4.0.0-rc.2, 
OS/X, node 6.9.5, npm 4.1.2

### Repro steps.
1) Open a CLI project with tests.
2) Develop tests in a single spec file with `ng test`, changing frequently.
3) Do the same again with `ng test --sourcemap=false` (soon to be `--sourcemaps=false`

Compare your experience.

### Observations

I believe that in the typical dev/test cycle, the programmer changes **a single test file** 80% of the time, the corresponding application file 10% of the time, and hardly anything else the remaining 10%.

My sense is that it would be wise to optimize for this workflow.

### PR DESCRIPTION
**Pull request by filipesilva**
This PR uses a new Karma plugin to enable vendor bundles in unit tests,
increasing rebuild performance.

On a medium size project rebuilds times were 15x smaller (16.5s to 0.9s).

Fix #5423

### CODE REVIEW

**Comment by hccampos** [comment]
Cool stuff! Any benchmarks you could share?

**Comment by filipesilva** [comment]
@hccampos I forgot to fill in the PR message with the commit message, it's updated now. TLDR 15x to 30x improvement, probably less for big projects.

**Comment by cexbrayat** [comment]
@filipesilva 👍 I just tried it, and it does start the test faster on rebuild. But to be honest, this does not completely fix #5423 (at least in my experiment): the progress report still go way slower on rebuild than on initial test:

```
Chrome 58.0.3029 (Mac OS X 10.12.4): Executed 102 of 102 SUCCESS (5.774 secs / 5.737 secs)
Chrome 58.0.3029 (Mac OS X 10.12.4): Executed 102 of 102 SUCCESS (28.755 secs / 28.73 secs)
Chrome 58.0.3029 (Mac OS X 10.12.4): Executed 102 of 102 SUCCESS (27.963 secs / 27.365 secs)
```

**Comment by filipesilva** [comment]
@cexbrayat bear in mind that we don't display how long **building** took, only how long karma took to run the tests. I haven't noticed delays while running the test though, so that's a relevant point.

Does this different in karma running time happen all the time, or only on some projects?

**Comment by cexbrayat** [comment]
@filipesilva Yes I know, I was just pointing out that (I think) #5423 has 2 parts. This PR solved the first one (rebuild took a long time, and it's now faster thanks to your PR 👍 ) but not the second part that @wardbell was pointing out: the progress report is ticking slower on rebuilds.
I'm witnessing this on pretty much all my projects.

**Comment by filipesilva** [comment]
@cexbrayat I wen't through a couple of rebuilds on my machine with both a brand new project and an existing medium sized project and couldn't notice any significant deviation on the reported karma re-test speeds (0.183/0.163/0.155 and 19.403/19.595/20.556).

In your case there's a whooping 500% difference though. Are you sure you're not experience chrome background throttling on your tests? I know chrome recently introduced that, and that it's been pretty bad for unit tests.

**Comment by mgol** [comment]
Good point @filipesilva. @cexbrayat make sure that during your testing the Chrome window is visible all the time, never covered by any other window.

If this was the problem, it will soon stop being an issue - in about a month Chrome 59 will be released with support for headless browsing and [Karma already supports that](https://github.com/karma-runner/karma-chrome-launcher/releases/tag/v2.1.0). For now you'd have to be using Chrome Canary or beta.

**Comment by cexbrayat** [comment]
@filipesilva @mgol wasn't the background throttling disabled in Karma since a while ago? But you're right, it does indeed speed my tests if my browser is not on the background 🤔 So don't mind my comment, I may be experiencing a Chrome/Mac issue (possibly related to https://bugs.chromium.org/p/chromium/issues/detail?id=605498).

**Comment by mgol** [comment]
@cexbrayat Chrome has added a `--disable-background-timer-throttling` flag and `karma-chrome-launcher` [uses it](https://github.com/karma-runner/karma-chrome-launcher/blob/7446181/index.js#L42) but, unfortunately, it doesn't fully unthrottle all the things.

The bug about that [has been open for more than a year](https://bugs.chromium.org/p/chromium/issues/detail?id=605498) but, alas, nothing seems to be done on this front.

**Comment by cexbrayat** [comment]
@mgol I gave a try to the `disable-renderer-backgrounding` flag and it does appear to solve my issue. Thx you and @filipesilva for putting me on right track 👍 

So I [opened a bug](https://github.com/karma-runner/karma-chrome-launcher/issues/123) on `karma-chrome-launcher` to have this flag included and [offered a PR](https://github.com/karma-runner/karma-chrome-launcher/pull/124). Let's see how it goes.

Meanwhile, for those experimenting the same issue than me, you can switch your `karma.conf.js` to:

```js
    customLaunchers: {
      chromeBackground: {
        base: 'Chrome',
        flags: ['--disable-renderer-backgrounding']
      }
    },
    browsers: ['chromeBackground'],
```

@filipesilva sorry for polluting this thread 🙏

**Comment by filipesilva** [comment]
@cexbrayat not at all, this is valuable info for people looking through this next time!

I wonder if we should add that renderer flag, but probably it's better to instead add a `--headless` flag for `ng test/e2e` when Chrome 59 comes out.

File: packages/@angular/cli/blueprints/ng/files/karma.conf.js [hunk 1]
```diff
@@ -15,25 +15,14 @@ module.exports = function (config) {
     client:{
       clearContext: false // leave Jasmine Spec Runner output visible in browser
     },
-    files: [
```
Comment (hansl) [inline packages/@angular/cli/blueprints/ng/files/karma.conf.js]: For projects using those, is this PR still working (even if it has the pre-PR performance hit)? Do we need to regenerate projects?

File: packages/@angular/cli/models/webpack-configs/test.ts [hunk 2]
```diff
@@ -49,17 +50,22 @@ export function getTestConfig(testConfig: WebpackTestOptions) {
...
+        name: 'inline'
+      }),
+      new webpack.optimize.CommonsChunkPlugin({
+        name: 'vendor',
+        chunks: ['main'],
+        minChunks: (module: any) => module.resource && module.resource.startsWith(nodeModules)
       }),
-      new KarmaWebpackEmitlessError()
-    ]
+      // new KarmaWebpackEmitlessError()
```
Comment (hansl) [inline packages/@angular/cli/models/webpack-configs/test.ts]: Forgotten comment.

File: packages/@angular/cli/blueprints/ng/files/karma.conf.js (hunk 1 above)
Comment (filipesilva) [inline packages/@angular/cli/blueprints/ng/files/karma.conf.js]: No need to regenerate projects. `tests/e2e/tests/test/test-backwards-compat.ts` tests that the old config work. 

The new plugin has some logic that removes these entries if they are present in the karma config, that ensures backwards compatibility.

File: packages/@angular/cli/models/webpack-configs/test.ts (hunk 2 above)
Comment (filipesilva) [inline packages/@angular/cli/models/webpack-configs/test.ts]: Fixed

**Comment by cexbrayat** [comment]
@filipesilva The PR I submitted to karma-chrome-launcher has been merged and a new version released: no need for us to tweak the karma.conf.js, we can only bump the version of the launcher.
I submitted a PR to do so (see #6216 )

**Comment by filipesilva** [comment]
@cexbrayat that's pretty cool, kudos!

**Pull request merged**